### DISCUSSION TITLE
How to handle 3rd party types not implementing IntoWasmAbi and FromWasmAbi?

### DISCUSSION THREAD

**Discussion by ChristianIvicevic** [wasm-bindgen/wasm-bindgen#3654] [Q&A]
For some precise computations I am using `rust_decimal::Decimal` in my structs as follows:

```rust
#[wasm_bindgen]
pub struct Precise {
    pub value: Decimal,
}
```

Of couse this doesn't compile as `Decimal` doesn't implement `IntoWasmAbi` and `FromWasmAbi`. Even manually attempting to implement them would not work since it's a type from a foreign crate. One workaround is a mapped struct that can be serialized across the border like this:

```rust
pub struct Precise {
    pub value: Decimal,
}

#[wasm_bindgen]
pub struct PreciseJs {
    pub value: f64,
}
```

But this introduces a lot of cognitive overhead doubling the amount of types just to be able to pass around data back and forth. Is there an idiomatic approach to that?

**Comment by daxpedda** [comment]
You could ask `rust_decimal` to implement `wasm-bindgen` on their types.

This generally seems like a similar problem to implementing Serde on foreign types, but `wasm-bindgen` doesn't have any nice ways to work around that AFAIK.

**Reply by ChristianIvicevic** [reply]
Do you know of any existing and possible small implementation in other crates? If it isn't too difficult I would consider contributing myself, but would need guidance to understand how to implement support for `wasm-bindgen` in general.

**Reply by daxpedda** [reply]
> Do you know of any existing and possible small implementation in other crates?

Unfortunately I am not.

>  If it isn't too difficult I would consider contributing myself, but would need guidance to understand how to implement support for `wasm-bindgen` in general.

I honestly have no experience exporting Rust to JS, only the other way around, but generally speaking you only have to add `#[wasm_bindgen]` at the top as long as all types are supported.
Additionally, there is an issue with exporting symbols from dependencies when the root package is a library itself, so using `#[wasm_bindgen]` in libraries will only work if the root package is a "binary". I didn't look too deep into this, but it seems to be a Rust issue to me.

**Comment by RedstoneWizard08** [comment]
What I did for my project was just forking the crate and adding it to the workspace, then manually patching it myself. I did this for `anyhow`, and it worked perfectly.

**Reply by ChristianIvicevic** [reply]
Yeah I am maintaining a fork for myself and have attached `#[wasm_bindgen]` to the relevant struct and methods. Even opened a PR but it seems the original maintainer isn't as active. Having to maintain a fork and possibly pull in upstream changes into it is always really annoying, but there isn't really any other way anyways.
